tests/out/quad-vert.msl

```diff
@@ -49,6 +51,6 @@ vertex main2Output main2(
     a_uv = a_uv1;
     a_pos = a_pos1;
     main1(v_uv, a_uv, _, a_pos);
-    const auto _tmp = type10 {v_uv, _.gl_Position, _.gl_PointSize, {_.gl_ClipDistance[0]}};
-    return main2Output { _tmp.member, _tmp.gl_Position1, _tmp.gl_PointSize1, {_tmp.gl_ClipDistance1[0]} };
+    const auto _tmp = type10 {v_uv, _.gl_Position, _.gl_PointSize, _.gl_ClipDistance};
+    return main2Output { _tmp.member, _tmp.gl_Position1, _tmp.gl_PointSize1, {_tmp.gl_ClipDistance1.inner[0]} };
```

Comment:
wait, so here we can't just do `_tmp.gl_ClipDistance1` for the last argument?

Reply:
Nope, unfortunately not. Not `gl_ClipDistance1.inner` either.

Reply:
hmm, I'm concerned now, a little bit. So when we are initializing the I/O struct, we can't just pass `_tmp.gl_ClipDistance1` because it's an array. But how would this work in other places? I.e. if you have a user function returning a user struct, then passing `{ xxx.inner[0], xxx.inner[1] }` would just fail horribly, since the only field of the target struct is `inner`.

Reply:
Afaik, only having a single pair of those `{` brackets when constructing a struct is okay if it only has one field. I think more testing would be nice, but I haven't found a (trivial) example that this breaks. Something like the following, for example, is fine:
```glsl
#version 450

vec3[8] hello_world() {
    vec3[8] xyz;
    return xyz;
}

struct X {
  vec3[8] yyy;
};

vec3[8] hmm2() {
  X x;
  x.yyy = hello_world();
  x.yyy[1] = vec3(3.0);
  return x.yyy;
}

layout(location = 0) out vec4 colour;

void main() {
    vec3[8] xyz = hmm2();
    colour = vec4(xyz[7], 1.0);
}
```
->
```metal
#include <metal_stdlib>
#include <simd/simd.h>

struct type3 {
    metal::float3 inner[8u];
};
struct X {
    type3 yyy;
};
constant metal::float3 const_type1_ = {3.0, 3.0, 3.0};

type3 hello_world(
) {
    type3 xyz;
    return xyz;
}

type3 hmm2_(
) {
    X x;
    type3 _e13 = hello_world();
    for(int _i=0; _i<8; ++_i) x.yyy.inner[_i] = _e13.inner[_i];
    x.yyy.inner[1] = const_type1_;
    return x.yyy;
}

void main1(
    thread metal::float4& colour
) {
    type3 xyz1;
    type3 _e13 = hmm2_();
    for(int _i=0; _i<8; ++_i) xyz1.inner[_i] = _e13.inner[_i];
    metal::float3 _e15 = xyz1.inner[7];
    colour = metal::float4(_e15.x, _e15.y, _e15.z, 1.0);
    return;
}

struct main2Output {
    metal::float4 member [[color(0)]];
};
fragment main2Output main2(
) {
    metal::float4 colour = {};
    main1(colour);
    return main2Output { colour };
}
```

Reply:
I mean, specifically for arrays.
```rust
struct Foo {
  xxx: array<f32; 3>;
};
fn bar() -> Foo {
  return Foo(array<f32;3>(1.0, 2.0, 3.0));
}
```
My understanding is that this PR right now is not going to handle this. Ideally, we'd want it to generate something like
```rust
const auto _tmp = Foo { .. };
return { { _tmp.xxx[0], _tmp.xxx[1], _tmp.xxx[2] } };
```
This will fail because the produced `Foo` would be declared with the `inner` in it, but the `return` code doesn't use it.

Reply:
```rust
struct Foo {
  xxx: array<f32, 3>;
};
fn bar() -> Foo {
  return Foo(array<f32, 3>(1.0, 2.0, 3.0));
}
```
fails for me with `error: invalid type cast to `Foo``. Looks like struct composition isn't implemented yet/ working?

Reply:
Sorry about this! Filed #792
Struct composition works fine for 2+ arguments:
```rust
struct Foo {
  xxx: array<f32, 3>;
  bar: i32;
};
fn bar() -> Foo {
  return Foo(array<f32, 3>(1.0, 2.0, 3.0), 2);
}
```

Reply:
```rust
struct Foo {
  xxx: array<f32, 3>;
  bar: i32;
};
fn bar() -> Foo {
  return Foo(array<f32, 3>(1.0, 2.0, 3.0), 2);
}

[[stage(vertex)]]
fn foo([[builtin(vertex_index)]] vi: u32) -> [[builtin(position)]] vec4<f32> {
	let foo = bar();
	return vec4<f32>(foo.xxx[0], foo.xxx[1], foo.xxx[2], 1.0);
}
```
translates to:
```metal
#include <metal_stdlib>
#include <simd/simd.h>

struct type1 {
    float inner[3];
};
struct Foo {
    type1 xxx;
    int bar;
};

Foo bar1(
) {
    return Foo {type1 {1.0, 2.0, 3.0}, 2};
}

struct fooInput {
};
struct fooOutput {
    metal::float4 member [[position]];
};
vertex fooOutput foo(
  metal::uint vi [[vertex_id]]
) {
    Foo _e1 = bar1();
    return fooOutput { metal::float4(_e1.xxx.inner[0], _e1.xxx.inner[1], _e1.xxx.inner[2], 1.0) };
}
```
and that compiles fine.

Reply:
Ok, I don't understand how `type1 {1.0, 2.0, 3.0}` part works :(
Type1 is a structure with 1 element. So shouldn't it yell at us for providing 3 elements instead of 1?

Reply:
Yeah I'm not 100% sure either, but what seems to happen is that (as I wrote in https://github.com/gfx-rs/naga/pull/764#discussion_r620511225):

> Afaik, only having a single pair of those `{` brackets when constructing a struct is okay if it only has one field

Reply:
Same thing is allowed in C:
```c
struct Arr {
    float inner[4];
};

int main() {
    struct Arr arr = { 1.0, 1.5, 2.0, 2.5 };
}
```

Reply:
Uh oh, in both MSL and C it lets you skip fields in the constructor though so
```metal
Foo bar1(
) {
    return Foo {type1 {1.0}, 2};
}
```
also works?? With no warnings or errors 😟 ??

Reply:
What's going to happen if the array only has one element? How is the compiler going to figure out if the initializer is for this field or for the `inner` itself? And initializers don't even have to be provided for all the fields. Ugh, C is weird :(